### PR TITLE
Allow HdfsAuthentication bindings to be changed

### PR DESCRIPTION
**Pull request by MiguelWeezardo**

### CODE REVIEW

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/authentication/UserGroupInformationUtils.java [hunk 1]
```diff
@@ -17,11 +17,11 @@
```
Comment (kokosing) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/authentication/UserGroupInformationUtils.java:17]: `Allow public access to UserGroupInformationUtils` > `Make UserGroupInformationUtils class public`

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/authentication/AuthenticationModules.java [hunk 2]
```diff
@@ -33,8 +34,8 @@ private AuthenticationModules() {}
 
     public static Module noHdfsAuthenticationModule()
     {
-        return binder -> binder
-                .bind(HdfsAuthentication.class)
+        return binder -> newOptionalBinder(binder, HdfsAuthentication.class)
+                .setDefault()
```
Comment (kokosing) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/authentication/AuthenticationModules.java:38]: I don't like optional binding. With optional binding it is hard to track what is the actual implementation. As one can override the binding. Maybe instead we could allow to replace the `HiveAuthenticationModule` with something else.

I need to think about it more.

**Comment by kokosing** [comment]
Also I think we could rename `HiveAuthenticationModule` to `HdfsAuthenticationModule`

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/authentication/AuthenticationModules.java (hunk 2 above)
Comment (MiguelWeezardo) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/authentication/AuthenticationModules.java:38]: The problem with module replacement is that the path by which `HiveAuthenticationModule` ends up being loaded in SEP is convoluted: `StarburstHivePlugin.getConnectorFactories()` -> `HiveConnectorFactory.create()` (reflection) -> `InternalHiveConnectorFactory.createConnector()` -> (a whole bunch of modules including `HiveAuthenticationModule`). Replacing just one module would require copying the class hierarchy: `StarburstHiveConnectorFactory.create()` -> `InternalStarburstHiveConnectorFactory.createConnector()` -> (the same bunch of modules with one replacement). Without serious refactoring on both sides, this would be a nightmare to keep in sync in case of Trino changes.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/InternalHiveConnectorFactory.java [hunk 3]
```diff
@@ -100,7 +100,7 @@ public static Connector createConnector(String catalogName, Map<String, String>
                     conditionalModule(RubixEnabledConfig.class, RubixEnabledConfig::isCacheEnabled, new RubixModule()),
                     new HiveMetastoreModule(metastore),
                     new HiveSecurityModule(catalogName),
-                    new HiveAuthenticationModule(),
+                    new HdfsAuthenticationModule(),
```
Comment (kokosing) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/InternalHiveConnectorFactory.java:103]: Class rename should be in separate commit

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/authentication/DirectHdfsAuthentication.java [hunk 4]
```diff
@@ -32,7 +33,7 @@ public DirectHdfsAuthentication(@ForHdfs HadoopAuthentication hadoopAuthenticati
     }
 
     @Override
-    public <R, E extends Exception> R doAs(String user, GenericExceptionAction<R, E> action)
+    public <R, E extends Exception> R doAs(ConnectorIdentity connectorIdentity, GenericExceptionAction<R, E> action)
```
Comment (kokosing) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/authentication/DirectHdfsAuthentication.java:36]: s/connectorIdentity/identity

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/authentication/HdfsAuthentication.java [hunk 5]
```diff
@@ -13,14 +13,16 @@
  */
 package io.trino.plugin.hive.authentication;
 
+import io.trino.spi.security.ConnectorIdentity;
+
 public interface HdfsAuthentication
 {
-    <R, E extends Exception> R doAs(String user, GenericExceptionAction<R, E> action)
+    <R, E extends Exception> R doAs(ConnectorIdentity connectorIdentity, GenericExceptionAction<R, E> action)
```
Comment (kokosing) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/authentication/HdfsAuthentication.java:20]: s/connectorIdentity/identity

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/authentication/HdfsAuthentication.java [hunk 6]
```diff
@@ -13,14 +13,16 @@
...
             throws E;
 
-    default void doAs(String user, Runnable action)
+    default void doAs(ConnectorIdentity connectorIdentity, Runnable action)
```
Comment (kokosing) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/authentication/HdfsAuthentication.java:23]: s/connectorIdentity/identity

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/authentication/ImpersonatingHdfsAuthentication.java [hunk 7]
```diff
@@ -33,10 +34,10 @@ public ImpersonatingHdfsAuthentication(@ForHdfs HadoopAuthentication hadoopAuthe
     }
 
     @Override
-    public <R, E extends Exception> R doAs(String user, GenericExceptionAction<R, E> action)
+    public <R, E extends Exception> R doAs(ConnectorIdentity connectorIdentity, GenericExceptionAction<R, E> action)
```
Comment (kokosing) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/authentication/ImpersonatingHdfsAuthentication.java:37]: s/connectorIdentity/identity

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/authentication/ImpersonatingHdfsAuthentication.java [hunk 8]
```diff
@@ -33,10 +34,10 @@ public ImpersonatingHdfsAuthentication(@ForHdfs HadoopAuthentication hadoopAuthe
...
             throws E
     {
-        return executeActionInDoAs(createProxyUser(user), action);
+        return executeActionInDoAs(createProxyUser(connectorIdentity.getUser()), action);
```
Comment (kokosing) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/authentication/ImpersonatingHdfsAuthentication.java:40]: You should provide a bean that handles how `identity` is transformed to `user`, so it could be customized.

**Review by kokosing — COMMENTED** [review]
The first commit should be no longer needed.

**Review by kokosing — APPROVED** [review]
Looks nice.

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/orc/TestOrcDeleteDeltaPageSource.java [hunk 9]
```diff
@@ -44,7 +45,7 @@ public void testReadingDeletedRows()
         File deleteDeltaFile = new File(Resources.getResource("fullacid_delete_delta_test/delete_delta_0000004_0000004_0000/bucket_00000").toURI());
         OrcDeleteDeltaPageSourceFactory pageSourceFactory = new OrcDeleteDeltaPageSourceFactory(
                 new OrcReaderOptions(),
-                "test",
+                ConnectorIdentity.forUser("test").build(),
```
Comment (kokosing) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/orc/TestOrcDeleteDeltaPageSource.java:48]: Use `ofUser`

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/orc/TestOrcDeletedRows.java [hunk 10]
```diff
@@ -154,15 +155,15 @@ private static OrcDeletedRows createOrcDeletedRows(AcidInfo acidInfo, String sou
         JobConf configuration = new JobConf(new Configuration(false));
         OrcDeleteDeltaPageSourceFactory pageSourceFactory = new OrcDeleteDeltaPageSourceFactory(
                 new OrcReaderOptions(),
-                "test",
+                ConnectorIdentity.forUser("test").build(),
```
Comment (kokosing) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/orc/TestOrcDeletedRows.java:158]: Use `ofUser`, same below

**Review by kokosing — COMMENTED** [review]
Please rebase.

**Review by findepi — COMMENTED** [review]
lgtm
